pkg/agent/monitortool/latency_store.go

```diff
@@ -0,0 +1,195 @@
+	// The status of the connection
+	Status bool
+	// The last time the connection was updated
+	LastUpdated time.Time
+	// The time the connection was created.
+	CreatedAt time.Time
+}
+
+func NewLatencyStore(nodeInformer coreinformers.NodeInformer) *LatencyStore {
+	store := &LatencyStore{
+		connectionMap: make(map[string]*Connection),
+		nodeInfoMap:   make(map[string]*corev1.Node),
+		nodeInformer:  nodeInformer,
+	}
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    store.OnNodeAdd,
+		UpdateFunc: store.OnNodeUpdate,
+		DeleteFunc: store.OnNodeDelete,
+	})
+
+	return store
+}
+
+func (l *LatencyStore) Run(stopCh <-chan struct{}) {
+	l.nodeInformer.Informer().Run(stopCh)
+}
+
+func (l *LatencyStore) OnNodeAdd(obj interface{}) {
+	node := obj.(*corev1.Node)
+	l.AddNodeToMap(node)
+}
+
+func (l *LatencyStore) OnNodeUpdate(oldObj, newObj interface{}) {
+	oldNode := oldObj.(*corev1.Node)
+	node := newObj.(*corev1.Node)
+	l.UpdateNodeByKey(oldNode.Name, node)
+}
+
+func (l *LatencyStore) OnNodeDelete(obj interface{}) {
+	node := obj.(*corev1.Node)
+	l.DeleteNodeByKey(node.Name)
+}
+
+func (l *LatencyStore) AddConnToMap(connKey string, conn *Connection) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	// Judge if the connection is already in the map
+	_, found := l.connectionMap[connKey]
+	if !found {
+		conn.CreatedAt = conn.LastUpdated
+	}
+
+	l.connectionMap[connKey] = conn
+}
+
+func (l *LatencyStore) GetConnByKey(connKey string) (*Connection, bool) {
+	l.mutex.RLock()
+	defer l.mutex.RUnlock()
+
+	conn, found := l.connectionMap[connKey]
+
+	return conn, found
+}
+
+func (l *LatencyStore) DeleteConnByKey(connKey string) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	delete(l.connectionMap, connKey)
+}
+
+func (l *LatencyStore) UpdateConnByKey(connKey string, conn *Connection) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	// Judge if the connection is already in the map
+	_, found := l.connectionMap[connKey]
+	if !found {
+		conn.CreatedAt = conn.LastUpdated
+	}
+
+	l.connectionMap[connKey] = conn
+}
+
+func (l *LatencyStore) ListConns() map[string]*Connection {
+	l.mutex.RLock()
+	defer l.mutex.RUnlock()
+
+	return l.connectionMap
+}
+
+func (l *LatencyStore) AddNodeToMap(node *corev1.Node) {
+	l.nodeInfoMap[node.Name] = node
+}
+
+func (l *LatencyStore) GetNodeByKey(nodeKey string) (*corev1.Node, bool) {
+	l.mutex.RLock()
+	defer l.mutex.RUnlock()
+
+	node, found := l.nodeInfoMap[nodeKey]
+
+	return node, found
+}
+
+func (l *LatencyStore) DeleteNodeByKey(nodeKey string) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	// Delete the node from the node info map
+	delete(l.nodeInfoMap, nodeKey)
+	// Delete the node from the connection map
+	delete(l.connectionMap, nodeKey)
+}
+
+func (l *LatencyStore) UpdateNodeByKey(nodeKey string, node *corev1.Node) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	l.nodeInfoMap[nodeKey] = node
+}
+
+func (l *LatencyStore) ListNodes() []corev1.Node {
```

Comment:
this function is only called by `ListNodesIPs` so I am not sure it makes sense to have it. We are just creating a temporary slice for no reason.